keeper/src/main/scala/zio/keeper/consensus/JGroups.scala

```diff
@@ -0,0 +1,147 @@
+            _       <- c.broadcast(payload)
+          } yield ()
+      }
+    }
+
+  private val msgLoop: ZIO[Consensus with ReceiverAdapter with Logging, Error, Unit] =
+    for {
+      c      <- ZIO.access[Consensus](_.get[Consensus.Service])
+      ra     <- ZIO.access[ReceiverAdapter](_.get[ReceiverAdapter.Service])
+      selfId <- c.selfNode
+      _ <- c.receive.tap { rawMsg =>
+            for {
+              msg <- TaggedCodec.read[JGroupsProtocol](rawMsg.payload)
+              _ <- msg match {
+                    case StateRequest =>
+                      for {
+                        _      <- logging.logInfo(s"[JGROUPS] Received state request")
+                        leader <- c.getLeader
+                        _ <- if (leader == selfId) {
+                              for {
+                                state   <- ra.getState
+                                payload <- TaggedCodec.write[JGroupsProtocol](StateResponse(state))
+                                _       <- c.send(payload, rawMsg.sender)
+                                _       <- logging.logInfo(s"[JGROUPS] Answering state request")
+                              } yield ()
+                            } else UIO.unit
+                      } yield ()
+                    case StateResponse(payload) =>
+                      for {
+                        _      <- logging.logInfo(s"[JGROUPS] Received state response")
+                        leader <- c.getLeader
+                        _ <- if (leader == rawMsg.sender) {
+                              ra.setState(payload)
+                            } else UIO.unit
+                      } yield ()
+                    case UserMessage(payload) =>
+                      logging.logInfo(s"[JGROUPS] Received user message") *> ra.receive(payload)
+                  }
+            } yield ()
+          }.runDrain
+    } yield ()
+
+  private val initializeState: ZIO[Clock with Logging with ReceiverAdapter with Consensus, Error, Boolean] =
+    for {
+      c       <- ZIO.access[Consensus](_.get[Consensus.Service])
+      ra      <- ZIO.access[ReceiverAdapter](_.get[ReceiverAdapter.Service])
+      leader  <- c.getLeader
+      payload <- TaggedCodec.write[JGroupsProtocol](StateRequest)
+      _       <- logging.logInfo(s"[JGROUPS] Asking for a state")
+      _       <- c.send(payload, leader)
+      response = c.receive
+        .filter(_.sender == leader)
+        .mapM(msg => TaggedCodec.read[JGroupsProtocol](msg.payload))
+        .collect { case s: StateResponse => s }
+        .run(ZSink.head[StateResponse])
+        .flatMap {
+          case None    => UIO(false)
+          case Some(s) => ra.setState(s.payload).map(_ => true)
+        }
+      timeout = clock.sleep(Duration.fromNanos(3000000000L)) *> UIO(false)
+      result  <- response.race(timeout)
+    } yield result
+
+  def create[T: Tagged](): ZLayer[Consensus with ReceiverAdapter with Clock with Logging with Has[T], Error, Has[T]] =
```

Comment:
No need for `()`, this is a pure function. *Nitpick:* start from `A` with type params.